### PR TITLE
Add keywords and specificators lecture materials

### PR DESCRIPTION
**Pull request by ivankochin**

### CODE REVIEW

File: 02_keywords_and_specificators/02_00_override_and_final/02_00_00_override_and_final.cpp [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+#include <iostream>
+
+struct base {
+    virtual void foo() { std::cout << "Base Foo" << std::endl;}
+    virtual void bar() { std::cout << "Base Bar" << std::endl;}
+};
+
+struct derived : public base {
+    void foo() /*const*/ override { std::cout << "Derived Foo" << std::endl;}
+    void bar() final { std::cout << "Derived Bar" << std::endl;}
```
Comment (ivankochin) [inline 02_keywords_and_specificators/02_00_override_and_final/02_00_00_override_and_final.cpp]:
```suggestion
    void bar() override final { std::cout << "Derived Bar" << std::endl; }
```

File: 02_keywords_and_specificators/02_00_override_and_final/02_00_00_override_and_final.cpp [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+struct derived_derived /*final*/: public derived {
+    void foo() override { std::cout << "Derived Derived Foo" << std::endl;}
+    void bar() override { std::cout << "Derived Derived Bar" << std::endl;}
+};
+
+struct derived_from_final : public derived_derived {};
+
+int main () {
+    derived_derived{}.foo();
```
Comment (ivankochin) [inline 02_keywords_and_specificators/02_00_override_and_final/02_00_00_override_and_final.cpp]: Add virtual call to the main function

File: 02_keywords_and_specificators/02_00_override_and_final/02_00_00_override_and_final.cpp [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+};
+
+struct derived : public base {
+    void foo() /*const*/ override { std::cout << "Derived Foo" << std::endl;}
+    void bar() final { std::cout << "Derived Bar" << std::endl;}
+};
+
+struct derived_derived /*final*/: public derived {
+    void foo() override { std::cout << "Derived Derived Foo" << std::endl;}
+    void bar() override { std::cout << "Derived Derived Bar" << std::endl;}
```
Comment (ivankochin) [inline 02_keywords_and_specificators/02_00_override_and_final/02_00_00_override_and_final.cpp]:
```suggestion
    // void bar() override { std::cout << "Derived Derived Bar" << std::endl; }
```

File: 02_keywords_and_specificators/02_00_override_and_final/02_00_00_override_and_final.cpp [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+// since C++11
+#include <iostream>
+
+struct base {
+    virtual void foo() { std::cout << "Base Foo" << std::endl;}
+    virtual void bar() { std::cout << "Base Bar" << std::endl;}
+};
+
+struct derived : public base {
+    void foo() /*const*/ override { std::cout << "Derived Foo" << std::endl;}
```
Comment (ivankochin) [inline 02_keywords_and_specificators/02_00_override_and_final/02_00_00_override_and_final.cpp]:
```suggestion
    void foo() /*const*/ override { std::cout << "Derived Foo" << std::endl; }
```

File: 02_keywords_and_specificators/02_00_override_and_final/02_00_00_override_and_final.cpp [hunk 5]
```diff
@@ -0,0 +1,25 @@
+// "override" and "final" keywords description
```
Comment (ivankochin) [inline 02_keywords_and_specificators/02_00_override_and_final/02_00_00_override_and_final.cpp]:
```suggestion
// "override" and "final" keywords behavior
```

File: 02_keywords_and_specificators/02_01_default_and_delete/02_01_00_default_and_delete.cpp [hunk 6]
```diff
@@ -0,0 +1,28 @@
+// "default" and "delete" keywords description
```
Comment (ivankochin) [inline 02_keywords_and_specificators/02_01_default_and_delete/02_01_00_default_and_delete.cpp]:
```suggestion
// "default" and "delete" keywords behavior
```

File: 02_keywords_and_specificators/02_01_default_and_delete/02_01_00_default_and_delete.cpp [hunk 7]
```diff
@@ -0,0 +1,28 @@
+// "default" and "delete" keywords description
+// which members may be "defaulted"
```
Comment (ivankochin) [inline 02_keywords_and_specificators/02_01_default_and_delete/02_01_00_default_and_delete.cpp]:
```suggestion
// allowed default members
```

File: 02_keywords_and_specificators/02_01_default_and_delete/02_01_00_default_and_delete.cpp [hunk 8]
```diff
@@ -0,0 +1,28 @@
+// "default" and "delete" keywords description
+// which members may be "defaulted"
+// note that "delete" declares the function but restricts to use it
```
Comment (ivankochin) [inline 02_keywords_and_specificators/02_01_default_and_delete/02_01_00_default_and_delete.cpp]:
```suggestion
// "delete" declares the function but restricts to use it
```

File: 02_keywords_and_specificators/02_02_friend/02_02_00_friend_declaration.cpp [hunk 9]
```diff
@@ -0,0 +1,23 @@
+#include <iostream>
```
Comment (ivankochin) [inline 02_keywords_and_specificators/02_02_friend/02_02_00_friend_declaration.cpp:5]:
```suggestion
// common use cases for friends (such as operators)
// friend definition inside the class
// friend always is not a member of the class

#include <iostream>
```

File: 02_keywords_and_specificators/02_02_friend/02_02_00_friend_declaration.cpp [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+};
+
+void my_function(const my_class& c) {
+    std::cout << c.field << std::endl;
+}
+
+template <typename T>
+struct my_struct {
+    T field = 0;
+    my_struct(const my_class& c) : field(c.field) {}
```
Comment (ivankochin) [inline 02_keywords_and_specificators/02_02_friend/02_02_00_friend_declaration.cpp]:
```suggestion
    my_struct(const my_class& c) : field(c.field) {}
private:
    int field;
```

File: 02_keywords_and_specificators/02_02_friend/02_02_00_friend_declaration.cpp [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+};
+
+int main () {
+    my_function(my_class{});
+
+    my_struct<int> s(my_class{});
```
Comment (ivankochin) [inline 02_keywords_and_specificators/02_02_friend/02_02_00_friend_declaration.cpp:27]:
```suggestion
    my_struct s(my_class{});
```

File: 02_keywords_and_specificators/02_02_friend/02_02_00_friend_declaration.cpp [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+    friend void my_function(const my_class&);
+    template<typename T> friend class my_struct;
+};
+
+void my_function(const my_class& c) {
+    std::cout << c.field << std::endl;
+}
+
+template <typename T>
+struct my_struct {
```
Comment (ivankochin) [inline 02_keywords_and_specificators/02_02_friend/02_02_00_friend_declaration.cpp:18]:
```suggestion
struct my_struct {
```

File: 02_keywords_and_specificators/02_02_friend/02_02_01_friend_definition.cpp [hunk 13]
```diff
@@ -0,0 +1,14 @@
+// A friend function may also be defined inside a class.
```
Comment (ivankochin) [inline 02_keywords_and_specificators/02_02_friend/02_02_01_friend_definition.cpp]: Remove this example

File: 02_keywords_and_specificators/02_04_auto/02_04_00_auto_definition.cpp [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+    auto& a_ref2 = i_ref;       // int&
+    auto& a_ref3 = const_i_ref; // const int&
+    auto& a_ref4 = i_ptr;       // int*&
+    // auto& a_ref0 = int{};    // compile-time error
+
+    const auto& const_a_ref0 = int{};       // const int&
+    const auto& const_a_ref1 = i;           // const int&
+    const auto& const_a_ref2 = i_ref;       // const int&
+    const auto& const_a_ref3 = const_i_ref; // const int&
+    const auto& const_a_ref4 = i_ptr;       // const int*&
```
Comment (ivankochin) [inline 02_keywords_and_specificators/02_04_auto/02_04_00_auto_definition.cpp]:
```suggestion
    auto a1 = 100;
    auto a1 = i;
    auto a2 = i_ref;
    auto a3 = const_i_ref;
    auto a4 = i_ptr;

    auto& a_ref1 = i;
    auto& a_ref2 = i_ref;
    auto& a_ref3 = const_i_ref;
    auto& a_ref4 = i_ptr;
    auto& a_ref0 = int{};

    const auto& const_a_ref0 = int{};
    const auto& const_a_ref1 = i;
    const auto& const_a_ref2 = i_ref;
    const auto& const_a_ref3 = const_i_ref;
    const auto& const_a_ref4 = i_ptr;
```

File: 02_keywords_and_specificators/02_04_auto/02_04_01_auto_return_type_deduction.cpp [hunk 15]
```diff
@@ -0,0 +1,25 @@
+// auto return type deduction
+
```
Comment (ivankochin) [inline 02_keywords_and_specificators/02_04_auto/02_04_01_auto_return_type_deduction.cpp]:
```suggestion
```

File: 02_keywords_and_specificators/02_04_auto/02_04_01_auto_return_type_deduction.cpp [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+// since C++14
+
+auto     void_func0() {}                  // returns void
+auto     void_func1() { return func0(); } // returns void
+// auto* void_func1() {}                  // compile-time error
+
+int x = 1;
+auto  func0() { return  x; } // return  int
+auto* func1() { return &x; } // returns int*
```
Comment (ivankochin) [inline 02_keywords_and_specificators/02_04_auto/02_04_01_auto_return_type_deduction.cpp]:
```suggestion
auto void_func0() {}                          // returns void
auto void_func1() { return func0(); } // returns void

int x = 1;
auto func0() { return  x; }    // return  int
// auto* void_func1() {}       // compile-time error
auto* func1() { return &x; } // returns int*
```

**Review by rarutyun — APPROVED** [review]

**Pull request merged**